src/api/bulk2.ts

```diff
@@ -0,0 +1,1082 @@
+    if (!options.pollTimeout) options.pollTimeout = this.pollTimeout;
+    if (!options.pollInterval) options.pollInterval = this.pollInterval;
+
+    const job = this.createJob({
+      object: options.object,
+      operation: options.operation,
+    });
+    try {
+      await job.open();
+      await job.uploadData(options.input);
+      await job.close();
+      await job.poll(options.pollInterval, options.pollTimeout);
+      return await job.getAllResults();
+    } catch (error) {
+      const err = error as Error;
+      this.logger.error(`bulk load failed due to: ${err}`);
+
+      if (err.name !== 'JobPollingTimeoutError') {
+        // fires off one last attempt to clean up and ignores the result | error
+        job.delete().catch((ignored) => ignored);
+      }
+      throw err;
+    }
+  }
+
+  /**
+   * Execute bulk query and get a record stream.
+   *
+   * Default timeout: 10000ms
+   *
+   * @param soql SOQL query
+   * @param BulkV2PollingOptions options object
+   *
+   * @returns {RecordStream} - Record stream, convertible to a CSV data stream
+   */
+  async query(
+    soql: string,
+    options?: Partial<BulkV2PollingOptions> & {
+      scanAll?: boolean;
+      columnDelimiter?: QueryJobInfoV2['columnDelimiter'];
+      lineEnding?: QueryJobInfoV2['lineEnding'];
+    },
+  ): Promise<Parsable<Record>> {
+    const queryJob = new QueryJobV2(this.connection, {
+      bodyParams: {
+        query: soql,
+        operation: options?.scanAll ? 'queryAll' : 'query',
+        columnDelimiter: options?.columnDelimiter,
+        lineEnding: options?.lineEnding,
+      },
+      pollingOptions: this,
+    });
+
+    const recordStream = new Parsable();
+    const dataStream = recordStream.stream('csv');
+
+    try {
+      await queryJob.open();
+      await queryJob.poll(options?.pollInterval, options?.pollTimeout);
+
+      const queryRecordsStream = await queryJob
+        .result()
+        .then((s) => s.stream());
+      queryRecordsStream.pipe(dataStream);
+    } catch (error) {
+      const err = error as Error;
+      this.logger.error(`bulk query failed due to: ${err}`);
+
+      if (err.name !== 'JobPollingTimeoutError') {
+        // fires off one last attempt to clean up and ignores the result | error
+        queryJob.delete().catch((ignored) => ignored);
+      }
+
+      recordStream.emit('error', err);
+      throw err;
+    }
+    return recordStream;
+  }
+}
+
+export class QueryJobV2<S extends Schema> extends EventEmitter {
+  private readonly connection: Connection<S>;
+  private readonly logger: Logger;
+  private readonly _id?: string;
+  private readonly bodyParams?: NewQueryJobOptions;
+  private readonly pollingOptions: BulkV2PollingOptions;
+  private error: Error | undefined;
+  private jobInfo?: QueryJobInfoV2;
+  private locator: Optional<string>;
+
+  constructor(conn: Connection<S>, options: ExistingQueryJobV2Options);
+  constructor(conn: Connection<S>, options: CreateQueryJobV2Options);
```

Comment:
overload constructor so that you provide either an `id` or `bodyParams` in options when creating a job instance.